powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/SignatureServiceBehavior.java

```diff
@@ -332,12 +334,38 @@ private ValidateSignatureResponse validateSignature(ActivationRecordEntity activ
         final PowerAuthSignatureTypes powerAuthSignatureTypes = signatureTypeConverter.convertFrom(signatureRequest.getSignatureType());
         List<SecretKey> signatureKeys = powerAuthServerKeyFactory.keysForSignatureType(powerAuthSignatureTypes, masterSecretKey);
 
+        // Get activation version and validate it
+        Integer version = activation.getVersion();
+        if (version == null || version < 2 || version > 3) {
+            throw localizationProvider.buildExceptionForCode(ServiceError.ACTIVATION_INCORRECT_STATE);
+        }
+
         // Verify the signature with given lookahead
         boolean signatureValid = false;
         long ctr = activation.getCounter();
         long lowestValidCounter = ctr;
+        byte[] ctrData = null;
+        HashBasedCounter hashBasedCounter = null;
+        // Get counter data from activation for version 3
+        if (version == 3) {
+            ctrData = BaseEncoding.base64().decode(activation.getCtrDataBase64());
+            hashBasedCounter = new HashBasedCounter();
+        }
+
         for (long iteratedCounter = ctr; iteratedCounter < ctr + powerAuthServiceConfiguration.getSignatureValidationLookahead(); iteratedCounter++) {
-            signatureValid = powerAuthServerSignature.verifySignatureForData(signatureRequest.getData(), signatureRequest.getSignature(), signatureKeys, iteratedCounter);
+            switch (version) {
+                case 2:
+                    // Use numeric counter for counter data
+                    ctrData = ByteBuffer.allocate(16).putLong(8, iteratedCounter).array();
+                    break;
+                case 3:
+                    // Increment counter in case iterated counter is bigger than current counter
+                    if (iteratedCounter > ctr) {
```

Comment:
This `if` seems a bit weird to me - I think we should separate the idea of "counter" (both numerical and hash-based) and "counter data" (data based on a counter value, used in the signature).

```java
ctrData = ctrHash; // use ctrHash data as ctrData
ctrHash = hashBasedCounter.next(ctrHash); // always increment the counter
```